File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int l1 = s1.length();  // Length of s1
        int l2 = s2.length();  // Length of s2

        // If s1 is longer than s2, permutation is not possible
        if (l1 > l2) {
            return false;
        }

        int[] charCount = new int[26]; // Frequency array for characters 'a' to 'z'

        // Initialize the frequency array for the first 'l1' characters in s1 and s2
        for (int i = 0; i < l1; i++) {
            charCount[s1.charAt(i) - 'a']--;  // Decrease count for s1
            charCount[s2.charAt(i) - 'a']++;  // Increase count for s2 (initial window)
        }

        // Count how many different characters exist in the window
        int diffCount = 0;
        for (int x : charCount) {
            if (x != 0) {
                diffCount++;
            }
        }

        // If no difference, then s1's permutation is found in s2
        if (diffCount == 0) {
            return true;
        }

        // Sliding window: shift window from left to right
        for (int i = l1; i < l2; i++) {
            int idx1 = s2.charAt(i - l1) - 'a'; // Old character (to be removed)
            int idx2 = s2.charAt(i) - 'a';      // New character (to be added)

            // Adding new character to the window
            if (charCount[idx2] == 0) {
                diffCount++; // If it was zero, we now introduce a difference
            }
            charCount[idx2]++; // Increase its count
            if (charCount[idx2] == 0) {
                diffCount--; // If it becomes zero again, we remove a difference
            }

            // Removing old character from the window
            if (charCount[idx1] == 0) {
                diffCount++; // If it was zero, we now introduce a difference
            }
            charCount[idx1]--; // Decrease its count
            if (charCount[idx1] == 0) {
                diffCount--; // If it becomes zero again, we remove a difference
            }

            // If all character differences are zero, we found a permutation
            if (diffCount == 0) {
                return true;
            }
        }

        // No permutation found
        return false;
    }
}



/*
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int l1 = s1.length();
        int l2 = s2.length();

        if(l1 > l2){
            return false;
        }

        int[] charCount = new int[26];

        for(int i = 0; i < l1; i++){
            charCount[s1.charAt(i)-'a']--;
            charCount[s2.charAt(i)-'a']++;
        }

        //count of diff char in s1 and s2
        int diffCount = 0;
        for(int x : charCount){
            if(x != 0){
                diffCount++;
            }
        }

        if(diffCount == 0){
            return true;
        }

        for(int i = l1; i < l2; i++){
            int idx1 = s2.charAt(i -l1)-'a'; //Old character to be removed from window
            int idx2 = s2.charAt(i)-'a'; //new character to be added from window

            //1--> Adding newChar to window
            // Before updating charCount for the new character check if its count is 0
            if(charCount[idx2] == 0){
                diffCount++;
            }
            // Add the new character as we move the window
            charCount[idx2]++;

            if(charCount[idx2] == 0){
                diffCount--;
            }

            //2--> Adding old char to window
            //check if its 0 then increment diffCount
            if(charCount[idx1] == 0){
                diffCount++;
            }
            // Remove the old character as we move the window
            charCount[idx1]--;

            if(charCount[idx1] == 0){
                diffCount--;
            }

            // If all deltas are zero, s1's permutation is found in s2.
            if (diffCount == 0) {
                return true;
            }
        }
        // If we reach here, no permutation of s1 is found in s2.
        return false;

    }
}
*/

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n1 = s1.length();
        int n2 = s2.length();

        // If s1 is longer than s2, no permutation can exist
        if (n1 > n2) {
            return false;
        }

        // Frequency arrays for s1 and the current window in s2
        vector<int> freq1(26, 0); // Frequency of characters in s1
        vector<int> freq2(26, 0); // Frequency of characters in the current window of s2

        // Initialize frequency counts for s1 and the first window of s2
        for (int i = 0; i < n1; i++) {
            freq1[s1[i] - 'a']++; // Increment count for s1
            freq2[s2[i] - 'a']++; // Increment count for the first window of s2
        }

        // Compare the frequency counts of the first window
        if (freq1 == freq2) {
            return true;
        }

        // Slide the window over s2
        for (int i = n1; i < n2; i++) {
            // Add the new character to the window
            freq2[s2[i] - 'a']++;
            // Remove the old character from the window
            freq2[s2[i - n1] - 'a']--;

            // Compare the frequency counts
            if (freq1 == freq2) {
                return true;
            }
        }

        // No matching window found
        return false;
    }
};

/*
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n1 = s1.size();
        int n2 = s2.size();

        if(n1 > n2){
            return false;
        }

        vector<int> s1Freq(26, 0);
        vector<int> s2Freq(26, 0);

        for(int i = 0; i < n1; i++){
            s1Freq[s1[i]-'a']++;
            s2Freq[s2[i]-'a']++;
        }

        if(s1Freq == s2Freq){
            return true;
        }


        for(int i = n1; i < n2; i++){
            s2Freq[s2[i]-'a']++;
            s2Freq[s2[i-n2]-'a']--;

            if(s1Freq == s2Freq){
                return true;
            }
        }

        // No matching window found
        return false;
    }
};
*/